### PR TITLE
RestJSONParser now parses "code" from "__type" field

### PR DESCRIPTION
**Pull request by rayluo**
According to issue boto/boto3#186, this happens in cloudsearchdomain service and some other services too. This fix is generic for all RestJSON responses.

### CODE REVIEW

File: botocore/parsers.py [hunk 1]
```diff
@@ -632,7 +632,7 @@ def _initial_body_parse(self, body_contents):
 
     def _do_error_parse(self, response, shape):
         body = self._initial_body_parse(response['body'])
-        error = {'Error': {}, 'ResponseMetadata': {}}
+        error = BaseJSONParser._do_error_parse(self, response, shape)
```
Comment (jamesls) [inline botocore/parsers.py]: Was there a reason not to use `super()`?

File: tests/unit/response_parsing/json/expected/elastictranscoder-delete-pipeline.json [hunk 2]
```diff
@@ -1,5 +1,5 @@
 {
-    "ResponseMetadata": {},
+    "ResponseMetadata": {"RequestId": "1234"},
```
Comment (jamesls) [inline tests/unit/response_parsing/json/expected/elastictranscoder-delete-pipeline.json]: Why was this added?

**Comment by jamesls** [comment]
Just had a couple questions, but overall looks good.

File: botocore/parsers.py (hunk 1 above)
Comment (rayluo) [inline botocore/parsers.py]: I should use super(). Will update the PR soon.

File: tests/unit/response_parsing/json/expected/elastictranscoder-delete-pipeline.json (hunk 2 above)
Comment (rayluo) [inline tests/unit/response_parsing/json/expected/elastictranscoder-delete-pipeline.json]: As a (harmless or even desirable?) side effect, RestJSONParser now inherits the parsing RequestId ability from the adjusted JSONParser.

I actually mentioned it in the commit log as below.

    Besides, RestJSONParser will now grab x-amzn-requestid header
    and populate the ResponseMetadata.
    An old JSON response test case is updated accordingly.

**Comment by jamesls** [comment]
Looks good :shipit:.  Please rebase and edit your commit message before merging.  Here's some guidelines that we follow: http://git-scm.com/docs/git-commit#_discussion, you can also look through the commit message for the CLI repo to get a feel for how the commit message are formatted.
